### PR TITLE
Fix None reading from ini files and more 

### LINKED ISSUE #26: Optional parameters without default should always accept None
.. as this is what is written out to config if they are not set.

### LINKED ISSUE #28: Crash on non-string invalid choice of parameter
If your your arguments are not of type string and you hand in a value not in the choices, then the error message that should be written crashes, because it is trying to make a string.

in dict_parser.py line 133

image

should be enough to remove the 's'

### LINKED ISSUE #29: Sphinx: Mocking numpy.random breaks scipy 1.4.1 imports
Should be checked if still needed, but most likely not.
Should just be removed from config.py in doc folder.

### LINKED ISSUE #30: Documentation improvements
- [x] add examples how to run script/call function 
  - ID_CONFIG = "entry_cfg"
  - ID_DICT = "entry_dict"
  - ID_JSON = "entry_json"
  - ID_SECTION = "section"

- [x] remove " (Quote) in "Example with list of dictionaries:"

### LINKED ISSUE #31: Improve create_parameter_help
- [x] improve formatting 
- [x] newlines after --Optional-- and --Required--

### LINKED ISSUE #34: Update Changelog for 1.0.7

### LINKED ISSUE #35: Update README
- [x] Remove JoschD from Authors

### PR DESCRIPTION
**Pull request by mihofer**
- [x] Fixes #26 
- [x] Fixes #28 
- [x] Fixes #29 
- [x] Fixes #30 
- [x] Fixes #31
- [x] Fixes #34  
- [x] Fixes #35

### CODE REVIEW

**Comment by mihofer** [comment]
Fixes #29 (?) .. apparently a non issue

**Comment by JoschD** [comment]
Fixes #30

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,4 +1,14 @@
 # Generic-Parser Changelog
+## Version 1.0.7
+ * fixed crash before proper error message for invalid choice of on non-string
```
Comment (fsoubelet) [inline CHANGELOG.md:3]: typo?

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,4 +1,14 @@
...
+ objects. 
+ * fixed `None` write and read in config files.
+ * added tests for list-lengths of default parameters.
+ * improved `create_parameter_help` function.
+ * some documentation fixed/additions.
+ * switched to github actions for CI.
```
Comment (fsoubelet) [inline CHANGELOG.md:9]: not no this branch?

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,4 +1,14 @@
...
+ * some code cleanup
```
Comment (fsoubelet) [inline CHANGELOG.md:10]: The user doesn't care

File: generic_parser/entrypoint_parser.py [hunk 4]
```diff
@@ -547,51 +589,71 @@ def add_to_arguments(args, entry_params=None, **kwargs):
...
     """
-    params = copy.deepcopy(params)
-    params = dict2list_param(params)
 
     if isinstance(parser, EntryPointParameters):
-        for param in params:
-            parser.add_parameter(**param)
+        parser = _add_params_to_entrypoint_parser(parser, params)
 
     elif isinstance(parser, ArgumentParser):
```
Comment (fsoubelet) [inline generic_parser/entrypoint_parser.py:602]: group with the previous statement?

File: CHANGELOG.md (hunk 2 above)
Comment (JoschD) [reply CHANGELOG.md:9]: true, but it's the Version 1.0.7 changelog ... this is all in the same project

File: CHANGELOG.md (hunk 1 above)
Comment (JoschD) [reply CHANGELOG.md:3]: fixed

File: CHANGELOG.md (hunk 3 above)
Comment (JoschD) [reply CHANGELOG.md:10]: true

File: CHANGELOG.md (hunk 2 above)
Comment (fsoubelet) [reply CHANGELOG.md:9]: Perfect

File: generic_parser/entrypoint_parser.py (hunk 4 above)
Comment (JoschD) [reply generic_parser/entrypoint_parser.py:602]: what?

Comment (fsoubelet) [reply generic_parser/entrypoint_parser.py:601]: The previous `if` and this `elif` lead to the exact same action.
Should be `if ... or ...`

File: tests/test_entrypoint.py [hunk 5]
```diff
@@ -147,27 +158,37 @@ def test_as_string():
...
+                  "--other"):
+        opt, unknown = paramtest_function()
+        assert opt.name == "myname"
+        assert opt.int == 3
+        assert len(opt.list) == 3
+        assert opt.list[1] == 5
+        assert len(unknown) > 0
+
+
+def test_as_config(tmp_output_dir):
```
Comment (mihofer) [inline tests/test_entrypoint.py:172]: since @fsoubelet made me suffer through it .. pytest tmp_path factory or similar would not be an option?

Comment (JoschD) [reply tests/test_entrypoint.py:172]: `tmp_path` is used in the fixture `tmp_output_dir` and is the standard option used if `DEBUG=False` (or not present)
see https://github.com/pylhc/generic_parser/blob/bug/26/NoneParameter/tests/conftest.py

Comment (mihofer) [reply tests/test_entrypoint.py:172]: oh, my bad, didnt see that ..

Comment (JoschD) [reply tests/test_entrypoint.py:172]: alternative way to do it (and use pytest only) would be to change the `basetmp` via config file. 
But as far as I can see from my tmp-folder, pytest keeps the symlinks but not the actual output after test-run?

File: tests/conftest.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    """
+    script = kwargs.get('script', 'somescript.py')
+    args_save = sys.argv.copy()
+    sys.argv = [script] + list(args)
+    yield
+    sys.argv = args_save
+
+
+@pytest.fixture
+def tmp_output_dir(tmp_path, request):
```
Comment (JoschD) [inline tests/conftest.py:36]: Just use `tmp_path` and the `--basetemp` option

**Review by JoschD — CHANGES_REQUESTED** [review]
Also check the line that wrapps strings into `""` in the save_config vs the line that should remove them in the loader  (write test explicitly)

File: tests/conftest.py (hunk 6 above)
Comment (JoschD) [reply tests/conftest.py]: done

**Review by JoschD — APPROVED** [review]

**Review by fsoubelet — APPROVED** [review]
Looks to me

**Pull request merged**